### PR TITLE
7685: Query-based specimen import

### PR DESCRIPTION
**Pull request by RosalineP**
#### Rationale
Allow for specimen import reload to be performed as an automated maintenance task, using data obtained from a schema, query, and optionally view, on Professional builds. Add UI and handling for undesired case of having two automated specimen importers active and loading for the same container.

#### Related Pull Requests
* https://github.com/LabKey/freezerpro/pull/23
* https://github.com/LabKey/professional/pull/2

#### Changes
- Adds ability to retrieve active importer from SpecimenService
- Adds interstitial page for configuring specimen import
- Changes SpecimenTransform's 'isEnabled()' to 'isValid()' and adds comments
- Does not allow a container's non-active specimen importer to run as a background task

Note: I misclicked and accidentally added @labkey-adam as a reviewer! Sorry Adam.

### CODE REVIEW

File: study/src/org/labkey/study/controllers/StudyController.java [hunk 1]
```diff
@@ -6212,6 +6210,61 @@ public ApiResponse execute(ReloadForm form, BindException errors) throws Excepti
...
+            return _activeTransform;
+        }
+
+        public void setActiveTransform(String activeTransform)
+        {
+            _activeTransform = activeTransform;
+        }
+    }
+
+    @RequiresPermission(AdminPermission.class) // TODO: What permission?
```
Comment (RosalineP) [inline study/src/org/labkey/study/controllers/StudyController.java]: Would you say this is the correct permission?

File: api/src/org/labkey/api/study/SpecimenService.java [hunk 2]
```diff
@@ -72,6 +72,8 @@ static SpecimenService get()
 
     Lsid getSpecimenMaterialLsid(@NotNull Container studyContainer, @NotNull String id);
 
+    String getActiveSpecimenImporter(@NotNull Container studyContainer);
```
Comment (labkey-klum) [inline api/src/org/labkey/api/study/SpecimenService.java:75]: add @Nullable annotation to indicate that this method can return a null value.

File: study/src/org/labkey/study/controllers/StudyController.java (hunk 1 above)
Comment (labkey-klum) [reply study/src/org/labkey/study/controllers/StudyController.java]: yes, Admin seems correct.

File: study/src/org/labkey/study/view/chooseImporter.jsp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+<labkey:errors/>
+
+<div>
+    <% if (numberOfTransforms > 1) { %>
+        <%
+            String selected = SpecimenService.get().getActiveSpecimenImporter(getContainer());
+        %>
+        <p>
+            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```
Comment (labkey-klum) [inline study/src/org/labkey/study/view/chooseImporter.jsp]: swap out placeholder text?

File: study/src/org/labkey/study/view/chooseImporter.jsp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+            </labkey:form>
+        </labkey:panel>
+    <% } else if (numberOfTransforms == 1) { %>
+        <%
+            SpecimenTransform transform = Iterables.get(specimenTransforms, 0);
+            ActionURL manageAction = transform.getManageAction(getContainer(), getUser());
+        %>
+        <p>
+            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```
Comment (labkey-klum) [inline study/src/org/labkey/study/view/chooseImporter.jsp]: fix placeholder text before we merge

**Review by labkey-klum — APPROVED** [review]
Just need to address the placeholder text in the JSP

File: api/src/org/labkey/api/study/SpecimenService.java (hunk 2 above)
Comment (RosalineP) [reply api/src/org/labkey/api/study/SpecimenService.java:75]: Done!

File: study/src/org/labkey/study/controllers/StudyController.java (hunk 1 above)
Comment (RosalineP) [reply study/src/org/labkey/study/controllers/StudyController.java]: Comment removed!

File: study/src/org/labkey/study/view/chooseImporter.jsp (hunk 4 above)
Comment (RosalineP) [reply study/src/org/labkey/study/view/chooseImporter.jsp]: Done 👍

File: study/src/org/labkey/study/view/chooseImporter.jsp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+            </labkey:form>
+        </labkey:panel>
+    <% } else if (numberOfTransforms == 1) { %>
+        <%
+            SpecimenTransform transform = Iterables.get(specimenTransforms, 0);
+            ActionURL manageAction = transform.getManageAction(getContainer(), getUser());
+        %>
+        <p>
+            Learn more about <%=helpLink("externalSpecimens", "Automated External Specimen Imports")%>.
```
Comment (RosalineP) [inline study/src/org/labkey/study/view/chooseImporter.jsp:110]: Might be a little sparse. The singleton importer page looks so weird, and from what I understand the majority of users would have one specimen importer at most, I'm leaning towards making the quick change of having the Manage tab link kick directly to the only available importer if there's only one. Thoughts?

**Review by RosalineP — COMMENTED** [review]
(Meant to click 'Comment' not 'Start Review', oops)

File: study/src/org/labkey/study/view/chooseImporter.jsp (hunk 3 above)
Comment (RosalineP) [reply study/src/org/labkey/study/view/chooseImporter.jsp]: Done

File: study/src/org/labkey/study/view/chooseImporter.jsp (hunk 5 above)
Comment (RosalineP) [reply study/src/org/labkey/study/view/chooseImporter.jsp:110]: Update: Implemented (and can take out) as of the commit 'NEW AND OPTIONAL: Support direct kick to only available Specimen Importer from manage tab' on both this PR and the FreezerPro PR.

**Pull request merged**